nemo/collections/nlp/parts/nlp_overrides.py

```diff
@@ -387,29 +388,33 @@ def save_checkpoint(
             assert (
                 len(checkpoint['optimizer_states']) == 1
             ), "Currently only support checkpointing 1 distributed optimizer per time!"
-            # converts the optimizer states to their sharded equivalents
-            sharded_optim_state = self.optimizer_sharded_state_dict(
-                unsharded_optim_state=checkpoint['optimizer_states'][0]
-            )
-
-            # Check whether to save optim states
-            include_optimizer = True if not storage_options else storage_options.get('include_optimizer', True)
-            if include_optimizer:
-                checkpoint['optimizer_states'] = [sharded_optim_state]
+            if self.save_top_k:
```

Comment:
I would suggest moving this logic out of the strategy, instead do the differentiation regular/.nemo save in the `save_checkpoint` method of the NeMoModelCallback. Otherwise there is so much entanglement in the logic.